Improve the Array article.

sideshowbarker opened the pull request:
`popularities.json` lists the `Array` article as one of our ten most-visited. So it merits some attention toward trying to make it as good a resource as possible for the many people coming by to read it. Thus this PR includes the following changes:

- streamlines the article intro down to just the essential points
- restructures the sections for a better information flow
- adds prose introductions to all the existing code examples
- gives all the examples linkable headings
- rewrites the code of many examples, for greater clarity/consistency
- makes various copy edits

sideshowbarker committed:
Improve the Array article.

This includes the following changes:

- streamlines the article intro down to just the essential points
- restructures the sections for a better information flow
- adds prose introductions to all the existing code examples
- gives all the examples linkable headings
- rewrites the code of many examples, for greater clarity/consistency
- makes various copy edits

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,36 +12,156 @@ browser-compat: javascript.builtins.Array
 {{JSRef}}
 
-The JavaScript **`Array`** class is a global object that is used in the construction of arrays; which are high-level, list-like objects.
+The **`Array`** object represents an [array](https://en.wikipedia.org/wiki/Array_data_structure), and has members for [performing common array operations](#examples).
```
FWIW, I would have a short but real definition here rather than link to array on wikipedia. I know that has lots of detail but as as an expert I won't bother, and as a beginner I'm already distracted. Just a line - "ordered collection of any type(s) of elements that can be iterated and indexed".

sideshowbarker replied:
Changed link target to https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays as we’d discussed

sideshowbarker committed:
Link to our own “JavaScript First Steps: Arrays” for “array” definition

sideshowbarker committed:
s/first/second/ typo for fruits[1] comment

sideshowbarker committed:
Re-order a couple of the Example sections

Re-order the “Call a function on each element in an array” section
and the “Access an array item by its index” section — to follow
“most-used first” order (basically).

- Create an array
- Access an array item by its index
- Find the index of an item in an array
- Append an item to an array
- Remove the last item from an array
- Remove the first item from an array
- Add a new first item to an array
- Remove a single item by index
- Remove multiple items by index
- Call a function on each element in an array
- Copy an array

ddbeck commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
 {{JSRef}}
 
-The JavaScript **`Array`** class is a global object that is used in the construction of arrays; which are high-level, list-like objects.
+The **`Array`** object represents an [array](/en-US/docs/Learn/JavaScript/First_steps/Arrays), and has members for [performing common array operations](#examples).
```
I like this rewrite overall, but I do think we should make some sort of effort to characterize an array in terms other than _array_. The rewrite does a good job of dumping a lot of accumulated… stuff, but I did like how the previous version characterized arrays in additional terms. I'd like to see something like that return here, even in a very limited way (e.g., "represents an array—a list-like sequence of values—and has members for…" would satisfy me, but I'm open to whatever).

hamishwillee replied:
@sideshowbarker Seems I'm not the only one with this comment :-). I know it was a "designed decision" but  the tautology "Array is an array" is hard to bare, and it wouldn't hurt to provide a few words (other than the more you say the more people will pick at). 

The learn link is good, so if it gets removed from here would be good to keep it as either further reading, or even as a link from the intro "if you need a softer introduction ..."

sideshowbarker replied:
OK, I went with this:

> The **`Array`** object, as with arrays in other programming languages, enables [storing a collection of multiple items under a single variable name](/en-US/docs/Learn/JavaScript/First_steps/Arrays), and has members for [performing common array operations](#examples).

ddbeck reviewed (commented):
Nice. One suggestion, but it's genuinely optional, since the rest is so good. 😄

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
+- **JavaScript arrays are [zero-indexed](https://en.wikipedia.org/wiki/Zero-based_numbering)**: the first element of an array is at index `0`, the second is at index `1`, and so on — and the last element is at the value of the array's {{jsxref("Array.length", "length")}} property minus `1`.
 
-**Create an Array**
+- **JavaScript [array-copy operations](#copy_an_array) create [shallow copies](/en-US/docs/Glossary/Shallow_copy)**, not [deep copies](/en-US/docs/Glossary/Shallow_copy) (just as standard built-in copy operations with _any_ JavaScript objects create shallow copies, not deep copies).
```
The deep copy link is wrong. I think below is just a little more readable.
```suggestion
- **JavaScript [array-copy operations](#copy_an_array) create [shallow copies](/en-US/docs/Glossary/Shallow_copy)**. (All standard built-in copy operations with _any_ JavaScript objects create shallow copies, rather than [deep copies](/en-US/docs/Glossary/Deep_copy)).
```

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
 ```
 
-**Remove items from an index position**
+### Remove multiple items by index
````
This is good, but there are other cool things that [`splice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) can do - some which you might mention here and others that you might mention elsewhere
- You can specify just start with no count value to strip off the end of the array from that point - I'd add something on that and maybe cross link from pop.
- You can use a negative start value. No idea how that works because the text in the splice docs is gibberish
  >  "If negative, it will begin that many elements from the end of the array. (In this case, the origin -1, meaning -n is the index of the nth last element, and is therefore equivalent to the index of array.length - n.) If start is negative infinity, it will begin from index 0.
- YOu can add elements at the same time as you remove them.

You might not want to add all these options, but could be good to add a line mentioning them.

sideshowbarker replied:
> This is good, but there are other cool things that [`splice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) can do - some which you might mention here and others that you might mention elsewhere
> 
>     * You can specify just start with no count value to strip off the end of the array from that point - I'd add something on that and maybe cross link from pop.

Added a **Truncate an array down to just its first _N_ items** example for that.

>     * You can use a negative start value. No idea how that works because the text in the splice docs is gibberish
>       > "If negative, it will begin that many elements from the end of the array. (In this case, the origin -1, meaning -n is the index of the nth last element, and is therefore equivalent to the index of array.length - n.) If start is negative infinity, it will begin from index 0.

Added a **Remove multiple items from the end of an array** example for that.

>     * YOu can add elements at the same time as you remove them.

Added a **Replace multiple items in an array** example for that.

hamishwillee replied:
This is all good. Thank you.

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
 
-This is a [shallow copy](/en-US/docs/Glossary/Shallow_copy) made using the [spread sequence](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) operator:
-Other ways to copy an array are discussed below in [Copying an array](#copying_an_array).
+All built-in array-copy operations ([spread syntax](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), [`Array.from()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from), [`Array.prototype.slice()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice), and [`Array.prototype.concat()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat)) create [ ...
```
First mention of `concat` - did you want to show that option too in example?

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
 
-This is a [shallow copy](/en-US/docs/Glossary/Shallow_copy) made using the [spread sequence](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) operator:
-Other ways to copy an array are discussed below in [Copying an array](#copying_an_array).
+All built-in array-copy operations ([spread syntax](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), [`Array.from()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from), [`Array.prototype.slice()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice), and [`Array.prototype.concat()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat)) create [ ...
```
I am _slightly_ concerned that we introduce only the stringify approach for deep copies, and that this is duplicating information in the deep copy. I was wondering if we might append "other approaches are covered in the deep copy topic". Something to think about.

sideshowbarker replied:
> I am _slightly_ concerned that we introduce only the stringify approach for deep copies, and that this is duplicating information in the deep copy. I was wondering if we might append "other approaches are covered in the deep copy topic". Something to think about.

Added a mention of `structuredClone()`.

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
-**Remove an item by index position**
+### Remove a single item by index
+
+This example uses the [`splice()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) method to remove the string "`Banana`" from the `fruits` array — by specifying the index position of "`Banana`".
```
It might be good to mention what happens to the removed element here or in the next topic. I think it is relevant.
```suggestion
This example uses the [`splice()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) method to remove the string "`Banana`" from the `fruits` array — by specifying the index position of "`Banana`".
Removed elements, in this case "`Banana`", are returned by the method in a new array.
```

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
+```
 
-JavaScript arrays are zero-indexed. The first element of an array is at index `0`, and the last element is at the index value equal to the value of the array's {{jsxref("Array.length", "length")}} property minus `1`.
+Finally, it’s important to understand that assigning an existing array to a new variable doesn’t create a copy. Instead the new variable is just a reference, or alias, to the original; that is, the original array’s name and the new variable name are just two names for the exact same object — and so, if you make any changes at all either to value of the original array or to the value of the new va ...
````
Maybe 
```suggestion
Finally, it’s important to understand that assigning an existing array to a new variable doesn’t create a copy of either the array or its elements. Instead the new variable is just a reference, or alias, to the original array; that is, the original array’s name and the new variable name are just two names for the exact same object — and so, if you make any changes at all either to value of the original array or to the value of the new variable, the other will change, too:
```

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
-Using an invalid index number returns `undefined`.
+```js
+let fruits = ['Strawberry', 'Mango']
+let fruitsAlias = fruits;
````
You might also show a test on true equality - isn't that `===` in JavaScript. Not advocating - throwing up for consideration. What you have done here shows the "point".

sideshowbarker replied:
> You might also show a test on true equality - isn't that `===` in JavaScript.

Added that.

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
+// ["Banana", "Strawberry"]
+```
+
+### Call a function on each element in an array
````
Should we also have an example on iterating through an array - this would be one way, but of course there are the others?

sideshowbarker replied:
> Should we also have an example on iterating through an array - this would be one way, but of course there are the others?

Added an **Iterate over an array** example, using `for...of` — with mention/links for the many other ways to iterate over any array.

hamishwillee replied:
Yes, this is just right.

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
+- **JavaScript arrays are resizable** and **can contain a mix of different [data types](/en-US/docs/Web/JavaScript/Data_structures)**. (When those characteristics are undesirable, use [typed arrays](/en-US/docs/Web/JavaScript/Typed_arrays) instead.)
 
-### Common operations
+- **JavaScript arrays are [zero-indexed](https://en.wikipedia.org/wiki/Zero-based_numbering)**: the first element of an array is at index `0`, the second is at index `1`, and so on — and the last element is at the value of the array's {{jsxref("Array.length", "length")}} property minus `1`.
```
I think here or maybe in another point we need to say that the elements are not associative arrays - they have to use numbers. Then link down to your notes.

sideshowbarker replied:
> I think here or maybe in another point we need to say that the elements are not associative arrays - they have to use numbers. Then link down to your notes.

👍 Added

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,377 @@ browser-compat: javascript.builtins.Array
+9  512  162
 ```
 
+### Notes
````
I think this should go up a level - it isn't an "other" example or really part of the previous example. If a more specific name than Notes could be used that would be awesome.

sideshowbarker replied:
>>  ### Notes 
> I think this should go up a level - it isn't an "other" example or really part of the previous example. If a more specific name than Notes could be used that would be awesome.

Thanks for catching that — I hadn’t actually realized I had it a level down; fixed now

sideshowbarker committed:
Update files/en-us/web/javascript/reference/global_objects/array/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

sideshowbarker committed:
Update files/en-us/web/javascript/reference/global_objects/array/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

sideshowbarker committed:
Update files/en-us/web/javascript/reference/global_objects/array/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
 
 ```js
+// 'fruits' array created using array literal notation.
 let fruits = ['Apple', 'Banana']
+console.log(fruits.length)
````
```suggestion
const fruits = ['Apple', 'Banana'];
console.log(fruits.length);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
+// 2
 
+// 'fruits' array created using the Array() constructor.
+let fruits = new Array('Apple', 'Banana');
 console.log(fruits.length)
```
```suggestion
const fruits = new Array('Apple', 'Banana');
console.log(fruits.length);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
 ```js
-let newLength = fruits.push('Orange')
-// ["Apple", "Banana", "Orange"]
+console.log(fruits.indexOf('Banana'))
````
```suggestion
console.log(fruits.indexOf('Banana'));
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
 ```js
-let last = fruits.pop() // remove Orange (from the end)
-// ["Apple", "Banana"]
+let newLength = fruits.push('Orange') // ["Apple", "Banana", "Orange"]
+console.log(newLength)
````
```suggestion
const newLength = fruits.push('Orange'); // ["Apple", "Banana", "Orange"]
console.log(newLength);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
-let first = fruits.shift() // remove Apple from the front
-// ["Banana"]
+// Remove 'Orange' from the end of 'fruits'.
+let last = fruits.pop() // ["Apple", "Banana"]
```
```suggestion
const last = fruits.pop(); // ["Apple", "Banana"]
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
-let newLength = fruits.unshift('Strawberry') // add to the front
-// ["Strawberry", "Banana"]
+// Remove 'Apple' from the front of 'fruits'.
+let first = fruits.shift() // ["Banana"]
```
```suggestion
const first = fruits.shift(); // ["Banana"]
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
-// 1
+```js
+// Add 'Strawberry' to the front of 'fruits'.
+let newLength = fruits.unshift('Strawberry') // ["Strawberry", "Banana"]
+console.log(newLength)
````
```suggestion
const newLength = fruits.unshift('Strawberry'); // ["Strawberry", "Banana"]
console.log(newLength);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
 ```js
-let removedItem = fruits.splice(pos, 1) // this is how to remove an item
+// This is how to remove a single item.
+let removedItem = fruits.splice(fruits.indexOf('Banana'), 1)
 
+console.log(fruits)
````
```suggestion
const removedItem = fruits.splice(fruits.indexOf('Banana'), 1);

console.log(fruits);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
+
+// fruits.splice(pos, 1) returns an array of the items
+// removed, which we assigned above to 'removedItem'.
+console.log(removedItem)
```
```suggestion
console.log(removedItem);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
-let vegetables = ['Cabbage', 'Turnip', 'Radish', 'Carrot']
-console.log(vegetables)
-// ["Cabbage", "Turnip", "Radish", "Carrot"]
+let fruits = ['Apple', 'Banana', 'Strawberry', 'Mango']
```
```suggestion
const fruits = ['Apple', 'Banana', 'Strawberry', 'Mango'];
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
+// This is how to remove multiple items: 'n' is the total
+// number of items to remove, starting at the index
+// position 'pos' and progressing toward the end of array.
+let removedItems = fruits.splice(pos, n)
```
```suggestion
const removedItems = fruits.splice(pos, n);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
 
-console.log(vegetables)
-// ["Cabbage", "Carrot"] (the original array is changed)
+console.log(fruits)
```
```suggestion
console.log(fruits);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
 
-### Accessing array elements
+```js
+let deepCopy = JSON.parse(JSON.stringify(fruits));
````
```suggestion
const deepCopy = (o) => JSON.parse(JSON.stringify(o));
const fruits2 = deepCopy(fruits);
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
 
-Using an invalid index number returns `undefined`.
+```js
+let fruits = ['Strawberry', 'Mango']
+let fruitsAlias = fruits;
````
```suggestion
const fruits = ['Strawberry', 'Mango'];
const fruitsAlias = fruits;
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
+let fruits = ['Strawberry', 'Mango']
+let fruitsAlias = fruits;
+// Any changes to the 'fruits' array change 'fruitsAlias' too.
+fruits.unshift('Apple', 'Banana')
```
```suggestion
fruits.unshift('Apple', 'Banana');
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
-console.log(arr[0])              // logs 'this is the first element'
-console.log(arr[1])              // logs 'this is the second element'
-console.log(arr[arr.length - 1]) // logs 'this is the last element'
+let board = [
+  ['R','N','B','Q','K','B','N','R'],
+  ['P','P','P','P','P','P','P','P'],
+  [' ',' ',' ',' ',' ',' ',' ',' '],
+  [' ',' ',' ',' ',' ',' ',' ',' '],
+  [' ',' ',' ',' ',' ',' ',' ',' '],
+  [' ',' ',' ',' ',' ',' ',' ',' '],
+  ['p','p','p','p','p','p','p','p'],
+  ['r','n','b','q','k','b','n','r'] ]
+
+console.log(board.join('\n') + '\n\n')
+
+// Move King's Pawn forward 2
+board[4][4] = board[6][4]
+board[6][4] = ' '
+console.log(board.join('\n'))
```
```suggestion
const board = [
  ['R','N','B','Q','K','B','N','R'],
  ['P','P','P','P','P','P','P','P'],
  [' ',' ',' ',' ',' ',' ',' ',' '],
  [' ',' ',' ',' ',' ',' ',' ',' '],
  [' ',' ',' ',' ',' ',' ',' ',' '],
  [' ',' ',' ',' ',' ',' ',' ',' '],
  ['p','p','p','p','p','p','p','p'],
  ['r','n','b','q','k','b','n','r'] ];

console.log(board.join('\n') + '\n\n');

// Move King's Pawn forward 2
board[4][4] = board[6][4];
board[6][4] = ' ';
console.log(board.join('\n'));
```

tshemsedinov commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
````diff
@@ -12,135 +12,378 @@ browser-compat: javascript.builtins.Array
+### Using an array to tabulate a set of values
+
+```js
+values = []
+for (let x = 0; x < 10; x++){
+ values.push([
+  2 ** x,
+  2 * x ** 2
+ ])
+}
+console.table(values)
````
```suggestion
const values = [];
for (let x = 0; x < 10; x++){
 values.push([
  2 ** x,
  2 * x ** 2,
 ]);
}
console.table(values);
```

tshemsedinov reviewed (commented):
Code style proposals

sideshowbarker committed:
Apply suggestions from code review

Co-authored-by: Timur Shemsedinov <timur.shemsedinov@gmail.com>

sideshowbarker committed:
Updates in response to review comments (more to come)

sideshowbarker committed:
Updates in response to review comments (more to come)

hamishwillee reviewed (approved):
Looks great. 

There are still a few of my suggestions inline - e.g. https://github.com/mdn/content/pull/12781/files#r801177204 and https://github.com/mdn/content/pull/12781/files#r801168231

I think they are worth you considering, but if you think they are not appropriate, this could be merged.

sideshowbarker commented:
> Looks great.
> 
> There are still a few of my suggestions inline - e.g. [#12781 (files)](https://github.com/mdn/content/pull/12781/files#r801177204) and [#12781 (files)](https://github.com/mdn/content/pull/12781/files#r801168231)

I plan today to make changes in response to both of those (and also for the `pop()` suggestion you noted elsewhere). Will ping you when I have it ready.

sideshowbarker committed:
Further updates in response to reviews comments, plus more examples

This change includes responses to review comments, plus the following
new examples:

- Create an array using String.prototype.split()
- Create a string from an array [using join()]
- Check if an array contains a certain item [using includes()]
- Truncate an array down to just its first _N_ items [using splice()]
- Replace multiple items in an array [using splice()]
- Iterate over an array [using for...of]
- Mention of various other ways to iterate over an array
- Merge multiple arrays together [using concat()]
- Mention of using structuredClone() to create deep copies

sideshowbarker commented:
> There are still a few of my suggestions inline - e.g. [#12781 (files)](https://github.com/mdn/content/pull/12781/files#r801177204) and [#12781 (files)](https://github.com/mdn/content/pull/12781/files#r801168231)
> 
> I think they are worth you considering, but if you think they are not appropriate, this could be merged.

Thanks much — I think I’ve now pushed changes to address all those.

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -12,444 +12,460 @@ browser-compat: javascript.builtins.Array
+- **JavaScript arrays are resizable** and **can contain a mix of different [data types](/en-US/docs/Web/JavaScript/Data_structures)**. (When those characteristics are undesirable, use [typed arrays](/en-US/docs/Web/JavaScript/Typed_arrays) instead.)
 
-### Common operations
+- **JavaScript arrays are not associative arrays** and so, [array elements cannot be accessed using strings as indexes](#notes) but must be accessed using integers as indexes.
```
```suggestion
- **JavaScript arrays are not associative arrays** and so, [array elements cannot be accessed using strings as indexes](#notes), but must be accessed using integers as indexes.
```

hamishwillee commented:
I just reviewed the changes that I mentioned. All look great. Still approved :-)

sideshowbarker committed:
Update files/en-us/web/javascript/reference/global_objects/array/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

ddbeck reviewed (approved):
Thank you, @sideshowbarker! I think this is going to make a real big difference for readers.  🎉

Alhadis commented on files/en-us/web/javascript/reference/global_objects/array/index.md:
```diff
@@ -503,6 +519,139 @@ Results in
 
+## Notes
+
+`Array` objects cannot use strings as element indexes (as in an [associative array](https://en.wikipedia.org/wiki/Associative_array)) but must use integers. Setting or accessing via non-integers using [bracket notation](/en-US/docs/Web/JavaScript/Guide/Working_with_Objects#objects_and_properties) (or [dot notation](/en-US/docs/Web/JavaScript/Reference/Operators/Property_Accessors)) will not set o ...
```
> `Array` objects cannot use strings as element indexes

This is wrong (or at the very least, poorly-worded). Arrays _can_ be indexed with strings, but only if they contain an (unsigned) integer ≥ 0:

~~~js
list = ["A", "B", "C"];
list["3"] = "D";
list.length === 4;
list.indexOf("D") === 3;
~~~

Conversely, an integer can be negative, whereas an array index cannot:

~~~js
list = [];
list[-1] = "Negative 1";
list.length === 0;
~~~
